shared/src/main/scala/mlscript/codegen/Codegen.scala

```diff
@@ -269,42 +270,34 @@ abstract class JSCode {
 }
 
 abstract class JSPattern extends JSCode {
-  def bindings: Ls[Str]
 }
 
 object JSPattern {
-  def bindings(patterns: Ls[JSPattern]): Ls[Str] = patterns.flatMap(_.bindings)
 }
 
 final case class JSArrayPattern(elements: Ls[JSPattern]) extends JSPattern {
-  def bindings: Ls[Str] = elements flatMap { _.bindings }
   def toSourceCode: SourceCode = SourceCode.array(elements map { _.toSourceCode })
 }
 
 final case class JSObjectPattern(properties: Ls[Str -> Opt[JSPattern]]) extends JSPattern {
   // If no sub-patterns, use the property name as the binding name.
-  def bindings: Ls[Str] = properties flatMap {
-    case name -> Some(subPattern) => subPattern.bindings
-    case name -> None             => name :: Nil
-  }
   def toSourceCode: SourceCode = SourceCode.record(
     properties
       .map {
-        case name -> Some(JSWildcardPattern()) => SourceCode(name)
+        case name -> Some(JSWildcardPattern()) => SourceCode(if (JSField.isValidIdentifier(name)) name else JSLit.makeStringLiteral(name))
         case name -> Some(subPattern) =>
-          SourceCode(s"$name: ") ++ subPattern.toSourceCode
-        case name -> N => SourceCode(name)
+          SourceCode(s"${if (JSField.isValidIdentifier(name)) name else JSLit.makeStringLiteral(name)}: ") ++ subPattern.toSourceCode
+        case name -> N => if (JSField.isValidIdentifier(name)) SourceCode(name)
+          else SourceCode(s"${JSLit.makeStringLiteral(name)}: ${Scope.replaceTicks(name)}")
```

Comment:
This name sanitization logic doesn't belong here at all. Here we should assume the names are fine. The `isValidIdentifier` function should be called whenever names are translated in the first place (such as in `declareParameter` AND in other places too).